### PR TITLE
[ACR-1579] Add machine-learning-training docs

### PR DESCRIPTION
**Pull request by zhoonit**
Ready to be reviewed.

### CODE REVIEW

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+   #include <nntrainer.h>
+   ```
+
+2. To use the Machine Learning Training API, include the following features in your `tizen-manifest.xml` file:
+
+   ```xml
+   <feature name="http://tizen.org/feature/machine_learning">true</feature>
+   <feature name="http://tizen.org/feature/machine_learning.training">true</feature>
+   ```
+
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:26]: add a statement about priveleges (if saving file outside the app, then apply privilege)?

Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:26]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 2]
```diff
@@ -0,0 +1,402 @@
+# Machine Learning Training
+
+Machine Learning Training allows to construct, control and train a machine learning model in Tizen devices.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:3]: Machine Learning Training API/system/functionality?

Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:3]: How about feature?

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+
+Please note that:
+
+* This feature remains experimental for Tizen 6.0 and is not full-fledged yet for every needs.
+* Every example code is not fully handling error for brevity. Error should be handled more extensively then example code written here.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:8]: not handling all the errors for brevity

Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:8]: then -> compared to the

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 4]
```diff
@@ -0,0 +1,402 @@
+# Machine Learning Training
+
+Machine Learning Training allows to construct, control and train a machine learning model in Tizen devices.
+
+Please note that:
+
+* This feature remains experimental for Tizen 6.0 and is not full-fledged yet for every needs.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:7]: Lets not add experimental. ACR reviewers were against it.
`This API is still undergoing heavy changes, and can be updated.`

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+* Every example code is not fully handling error for brevity. Error should be handled more extensively then example code written here.
+
+## Prerequisites
+
+To enable your application to use the machine learning training functionality:
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:12]: follow as per https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+1. To use the functions and data types of the Machine Learning Inference API, include the `<nntrainer.h>` header file in your application:
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Inference -> Training

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 4 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:7]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 3 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:8]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 6 above)
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:14]: API -> as per  https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 5 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:12]: Did you mean this? 
```suggestion
To enable your application to use the machine learning training functionality, follow as per:
```

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 6 above)
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Simply `include the `<nntrainer.h>` header file in your application:`. The first part of the sentence is already at line or modify line 12 to have it,.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+
+   ```c
+   #include <nntrainer.h>
+   ```
+
+2. To use the Machine Learning Training API, include the following features in your `tizen-manifest.xml` file:
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:20]: ditto

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+
+   ```xml
+   <feature name="http://tizen.org/feature/machine_learning">true</feature>
+   <feature name="http://tizen.org/feature/machine_learning.training">true</feature>
+   ```
+
+3. If you want to save/load model files from outside, the application has to request permission by adding following privileges to the `tizen-manifest.xml` file:
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:27]: outside `of the applicaiton's own resources`

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+    <!-- For accessing media storage -->
+    <privilege>http://tizen.org/privilege/mediastorage</privilege>
+    <!-- For accessing external storage -->
+    <privilege>http://tizen.org/privilege/externalstorage</privilege>
+  </privileges>
+  ```
+
+## Building Blocks
+
+There are 4 major components you need to get familiar with. With those components, you can prepare a model and run.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:40]: run?? train etc?

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 8 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:27]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 7 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:20]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 6 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:14]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 9 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:40]: Changed to 
```suggestion
There are 4 major components you need to get familiar with. With those components, you can train a model.
```

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 5 above)
Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:12]: I mean machine learning training `functionality` -> as per decided from https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 2 above)
Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:3]: this will get mixed with Tizen features ().
Further, you use API at bottom. Make sure whatever you choose works with it.

Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:3]: Make it API, consistent wtih inference.

Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:3]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+
+### Model
+
+![model](./media/model.png)
+
+Model is a wrapper component that has topology of layers, optimizers, datasets.
+Model conducts training and update parameters that can later be used for inference.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:47]: Model performs training and saves the updated parameters ....

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 5 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:12]: Changed to API

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+Please note only sequential topology is supported for the time being.
+
+```c
+// Create model
+ml_train_model_h model;
+ml_train_model_construct(&model);
+
+/* Configure models(omitted for brevity) */
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:55]: models -> model

Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:55]: omitted for brevity -> detailed in later sections

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 10 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:47]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 11 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:55]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+
+// after add layer to model,
+// you do not need to destroy layer since ownership is transferred to the model.
+ml_train_model_add_layer(model, layer);
+```
+
+### Optimizer
+
+Optimizer determines how to update model parameters according to loss from prediction.
+Currently, Stochastic gradient descent optimizer and Adam optimizer are supported.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:93]: Stochastic Gradient Descent

Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:93]: Currently -> Both

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+ml_train_dataset_set_property(dataset, "buffer_size=100", NULL);
+
+// after set dataset to model,
+// you do not need to destroy dataset since ownership is transferred to the model.
+ml_train_model_set_dataset(model, dataset);
+```
+
+## Construct Model
+
+To run the model, you need to construct model. Model can be constructed with `ml_train_model_construct`.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:129]: you need to construct -> a model first needs to be construct.
Do similar across the content of the file.

Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:129]: run the model -> what does run means - inference or train?

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+If you have file that describe model, the file can be used to construct initially with `ml_train_model_construct_with_file`.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:130]: describe -> describes

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+Even if you constructed model from a file, you can switch/modify/add each component with set of given apis until you compile with `ml_train_model_compile`.
+
+### Construct Model from the Description File
+
+As of now, only `*.ini` is supported to construct model from a file.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:135]: As of now, only INI formatted files `*.ini` are supported

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+
+#### Create Model from ini Format
+
+There are special sections refering to `model` and `dataset` respectively [Model] and [Dataset]
+Each ini section maps to a `layer`. Keys and values from each section sets property.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:140]: Rest of the ini sections map to `layer`.

Comment (kparichay) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:140]: Also add each layer name to unique.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 13 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:129]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 14 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:130]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 15 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:135]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 16 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:140]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 12 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:93]: Done

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+Invalid Property will be ignored safely, invalid values will cause error. All keys and values are case-insensitive
+
+For example,
+
+```ini
+[Model] # Special section that describes model property
+Type = NeuralNetwork	# Model Type : only NeuralNetwork is supported as of now
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:147]: such information `only NeuralNetwork is supported as of now` must not be given here. But rather in docs of the project.

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+# Layer Sections, each section name refers to name of the layer
+[inputlayer]
+Type = input
+Input_Shape = 1:1:62720	# Input Layer Dimension channel:height:width
+Normalization = true
+
+[outputlayer]
+Type = fully_connected
+Unit = 2		# Output Layer Dimension ( = Weight Width )
+bias_initializer = zeros
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:182]: lets specify weight initializer as well for completeness

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+Activation = sigmoid 	# activation : sigmoid, softmax
+Weight_Decay = l2norm
+weight_Decay_Lambda = 0.005
+```
+
+There are some restriction that must be kept:
+
+ * Model file must have `[Model]` section that describe model property
+ * Model file must have at lease one layer.
```
Comment (kparichay) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:191]: at least

**Review by kparichay — COMMENTED** [review]
I will review the rest and update this directly in the next PR.

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 2 above)
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:3]: I think we need to give little bit more details (briefly) about Machine Learning Training.
Something like..

Machine Learning Training API allows you to train your model in Tizen Devices.

The main features of the Machine Learning Training API include:

Construct deep neural network Easily
You can construct DNN model by Configuration File or writing code through Machine Learning Training API.
....
Train with your own data
Machine Learning API also allows you to train the model with your own data as a File I/O and defining data generator code.
....
Evaluate the model during train
You can validate and test your model during train process easily by define the data set.
...

File: docs/application/native/guides/machine-learning/machine-learning-training.md [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+ml_train_model_destroy(model);
+```
+
+### Layer
+
+Layer is a component that does actual computation while managing internal trainable parameters.
+
+```
+/***************************************
+ * Data(x)  -> Layer(F) -> Data(F(X))  *
```
Comment (jijoongmoon) [inline docs/application/native/guides/machine-learning/machine-learning-training.md:72]: I think we don't need this. I think we need to give info about what kind of layer we can provide.

Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:72]: Deleted

File: docs/application/native/guides/machine-learning/machine-learning-training.md (hunk 2 above)
Comment (zhoonit) [reply docs/application/native/guides/machine-learning/machine-learning-training.md:3]: Updated
